### PR TITLE
allow to add dependencies and php extensions into the Nextcloud container

### PR DESCRIPTION
**Pull request by szaimen**
Close https://github.com/nextcloud/all-in-one/issues/1162

Signed-off-by: Simon L <szaimen@e.mail.de>

### CODE REVIEW

**Comment by szaimen** [comment]
@Zoey2936 WDYT? :)

File: Containers/nextcloud/start.sh [hunk 1]
```diff
@@ -36,6 +36,34 @@ if ! [ -f "$NEXTCLOUD_DATA_DIR/this-is-a-test-file" ]; then
 fi
 sudo -u www-data rm -f "$NEXTCLOUD_DATA_DIR/this-is-a-test-file"
 
+# Install additional dependencies
+if [ -n "$ADDITIONAL_APKS" ]; then
+    if ! [ -f "/additional-apks-are-installed" ]; then
+        read -ra ADDITIONAL_APKS_ARRAY <<< "$ADDITIONAL_APKS"
+        for app in "${ADDITIONAL_APKS_ARRAY[@]}"; do
+            apk add "$app"
```
Comment (Zoey2936) [inline Containers/nextcloud/start.sh:44]: Why not also:
```suggestion
            apk add --no-cache "$app"
```

**Review by Zoey2936 — COMMENTED** [review]
I think it should work, but I think it would also be good to have an option to install packages using
```
npm install --global <package> - requieres to have npm installed using apk
yarn global add <package> - requieres to have yarn installed using apk
pip install <packages> - requieres to have py3-pip installed using apk
```
Because for example the NCDownloader app requires installing yt-dlp using pip (which requires gcc, musl-dev and py3-pip to be installed using apk)
I currently don't know if any npm/yarn packages is needed by any nextcloud app existing, but I think at some point in the future it could be also needed

**Comment by szaimen** [comment]
> Because for example the NCDownloader app requires installing yt-dlp using pip (which requires gcc, musl-dev and py3-pip to be installed using apk)

https://pkgs.alpinelinux.org/packages?name=youtube-dl&branch=v3.16&repo=&arch=&maintainer=

**Comment by Zoey2936** [comment]
> > Because for example the NCDownloader app requires installing yt-dlp using pip (which requires gcc, musl-dev and py3-pip to be installed using apk)
> 
> https://pkgs.alpinelinux.org/packages?name=youtube-dl&branch=v3.16&repo=&arch=&maintainer=

There is a difference between youtube-dl and yt-dlp

**Comment by szaimen** [comment]
> There is a difference between [youtube-dl](https://github.com/ytdl-org/youtube-dl) and [yt-dlp](https://github.com/yt-dlp/yt-dlp)
> And [NCDownloader referces to yt-dlp](https://github.com/shiningw/ncdownloader#installing-aria2-and-yt-dlp-in-ubuntu)

https://pkgs.alpinelinux.org/packages?name=yt-dlp&branch=v3.16&repo=&arch=&maintainer=

**Comment by Zoey2936** [comment]
Then apk packages and php extension should be enough, then there is only the `--no-cache` question, if it is needed

File: Containers/nextcloud/start.sh (hunk 1 above)
Comment (szaimen) [reply Containers/nextcloud/start.sh:44]: sounds good, will apply that later!

File: Containers/nextcloud/start.sh [hunk 2]
```diff
@@ -36,6 +36,34 @@ if ! [ -f "$NEXTCLOUD_DATA_DIR/this-is-a-test-file" ]; then
...
+if [ -n "$ADDITIONAL_PHP_EXTENSIONS" ]; then
+    if ! [ -f "/additional-php-extensions-are-installed" ]; then
+        read -ra ADDITIONAL_PHP_EXTENSIONS_ARRAY <<< "$ADDITIONAL_PHP_EXTENSIONS"
+        for app in "${ADDITIONAL_PHP_EXTENSIONS_ARRAY[@]}"; do
+            if [ "$app" = imagick ]; then
+                pecl install imagick-3.7.0
+                docker-php-ext-enable imagick
+            else
+                pecl install "$app"
+                docker-php-ext-enable "$app"
```
Comment (szaimen) [inline Containers/nextcloud/start.sh:60]: It seems like that some extensions are not on pecl but instead delivered via core. If pecl fails, we need to run something like https://github.com/nextcloud/all-in-one/blob/b027cdd2933ede2f072fb4259147d32cc0276572/Containers/nextcloud/Dockerfile#L49 instead

Comment (szaimen) [reply Containers/nextcloud/start.sh:60]: or maybe try to install it via core and afterwards via pecl. Need to figure out the best order.

Comment (szaimen) [reply Containers/nextcloud/start.sh:60]: I guess best is to check if part of core by looking into `/usr/src/php/ext/`. If yes, install using docker-php-ext-install, if not, install using pecl.

File: docker-compose.yml [hunk 3]
```diff
@@ -28,6 +28,8 @@ services:
       # - NEXTCLOUD_MAX_TIME=3600 # Can be adjusted if you need more. See https://github.com/nextcloud/all-in-one#how-to-adjust-the-max-execution-time-for-nextcloud
       # - TRUSTED_CACERTS_DIR=/path/to/my/cacerts # CA certificates in this directory will be trusted by the OS of the nexcloud container (Useful e.g. for LDAPS) See See https://github.com/nextcloud/all-in-one#how-to-trust-user-defiend-certification-authorities-ca
       # - COLLABORA_SECCOMP_DISABLED=false # Setting this to true allows to disable Collabora's Seccomp feature. See https://github.com/nextcloud/all-in-one#how-to-disable-collaboras-seccomp-feature
+      # - NEXTCLOUD_ADDITIONAL_APKS=imagick # This allows to add additional packages to the Nextcloud container permanently.
+      # - NEXTCLOUD_ADDITIONAL_PHP_EXTENSIONS=imagick # This allows to add additional php extensions to the Nextcloud container permanently. Default is imagick but can be overwritten by modifying this value.
```
Comment (szaimen) [inline docker-compose.yml:32]: we coudl add a link to the readme here

File: Containers/nextcloud/start.sh [hunk 4]
```diff
@@ -36,6 +36,34 @@ if ! [ -f "$NEXTCLOUD_DATA_DIR/this-is-a-test-file" ]; then
...
+fi
+
+# Install additional php extensions
+if [ -n "$ADDITIONAL_PHP_EXTENSIONS" ]; then
+    if ! [ -f "/additional-php-extensions-are-installed" ]; then
+        read -ra ADDITIONAL_PHP_EXTENSIONS_ARRAY <<< "$ADDITIONAL_PHP_EXTENSIONS"
+        for app in "${ADDITIONAL_PHP_EXTENSIONS_ARRAY[@]}"; do
+            if [ "$app" = imagick ]; then
+                pecl install imagick-3.7.0
+                docker-php-ext-enable imagick
```
Comment (szaimen) [inline Containers/nextcloud/start.sh:57]:
```suggestion
                echo "Installing Imagick via PECL..."
                pecl install imagick-3.7.0 >/dev/null
                docker-php-ext-enable imagick >/dev/null
```

File: Containers/nextcloud/start.sh (hunk 2 above)
Comment (szaimen) [inline Containers/nextcloud/start.sh:60]:
```suggestion
                echo "Installing $app via PECL..."
                pecl install "$app" >/dev/null
                docker-php-ext-enable "$app" >/dev/null
```

File: Containers/nextcloud/start.sh (hunk 1 above)
Comment (szaimen) [reply Containers/nextcloud/start.sh:44]:
```suggestion
            echo "Installing $app via apk..."
            apk add --no-cache "$app" >/dev/null
```

File: .github/workflows/nextcloud-update.yml [hunk 5]
```diff
@@ -57,8 +57,8 @@ jobs:
             | sort -V \
             | tail -1
         )"
-        sed -i "s|pecl install imagick.*\;|pecl install imagick-$imagick_version\;|" ./Containers/nextcloud/Dockerfile
-        
+        sed -i "s|pecl install imagick.*|pecl install imagick-$imagick_version|" ./Containers/nextcloud/start.sh
```
Comment (szaimen) [inline .github/workflows/nextcloud-update.yml:60]:
```suggestion
        sed -i "s|pecl install imagick.*|pecl install imagick-$imagick_version  >/dev/null|" ./Containers/nextcloud/start.sh
```

File: php/src/Docker/DockerActionManager.php [hunk 6]
```diff
@@ -332,6 +332,10 @@ public function CreateContainer(Container $container) : void {
                     $replacements[1] = $this->configurationManager->GetCollaboraSeccompPolicy();
                 } elseif ($out[1] === '%NEXTCLOUD_STARTUP_APPS%') {
                     $replacements[1] = $this->configurationManager->GetNextcloudStartupApps();
+                } elseif ($out[1] === '%NEXTCLOUD_ADDITIONAL_APKS%') {
+                    $replacements[1] = $this->configurationManager->GetNextcloudAdditionalApks();
+                } elseif ($out[1] === '%NEXTCLOUD_ADDITIONAL_PHP_EXTENSIONS%') {
```
Comment (szaimen) [inline php/src/Docker/DockerActionManager.php:337]:
```suggestion
                } elseif ($out[1] === 'NEXTCLOUD_ADDITIONAL_APKS') {
                    $replacements[1] = $this->configurationManager->GetNextcloudAdditionalApks();
                } elseif ($out[1] === 'NEXTCLOUD_ADDITIONAL_PHP_EXTENSIONS') {
```

File: readme.md [hunk 7]
```diff
@@ -467,6 +467,16 @@ You can run AIO also with docker rootless. How to do this is documented here: [d
...
 
+### How to add packets permanently to the Nextcloud container?
+Some Nextcloud apps require additional external dependencies that must be bundled within Nextcloud container in order to work correctly. As we cannot put each and every dependency for all apps into the container - as this would make the project very fast unmaintainable - there is an official way how you can add additional dependencies into the Nextcloud container. However note that doing this is not recommended since we do not test Nextcloud apps that require external dependencies. 
+
+You can do so by adding `-e NEXTCLOUD_ADDITIONAL_APKS=dependency1 dependency2` to the docker run command of the mastercontainer and customize the value to your fitting. It must be a string with small letters a-z, spaces and hyphens or '_'. You can find available packages here: https://pkgs.alpinelinux.org/packages?name=&branch=v3.16&repo=&arch=&maintainer=
```
Comment (szaimen) [inline readme.md:473]:
```suggestion
You can do so by adding `-e NEXTCLOUD_ADDITIONAL_APKS="dependency1 dependency2"` to the docker run command of the mastercontainer and customize the value to your fitting. It must be a string with small letters a-z, spaces and hyphens or '_'. You can find available packages here: https://pkgs.alpinelinux.org/packages?name=&branch=v3.16&repo=&arch=&maintainer=
```

File: Containers/nextcloud/start.sh (hunk 2 above)
Comment (szaimen) [reply Containers/nextcloud/start.sh]: okay, just tested and it is not possible to check this beforehand based on a folder.

Comment (szaimen) [reply Containers/nextcloud/start.sh]: found a solution

**Review by Zoey2936 — APPROVED** [review]
Looks good

**Comment by szaimen** [comment]
Thanks Zoey! :)

**Comment by Zoey2936** [comment]
I think numbers should also be allowed
grafik

**Comment by szaimen** [comment]
good catch. Let me add this...

**Comment by szaimen** [comment]
to review: https://github.com/nextcloud/all-in-one/pull/1387

**Pull request merged**